llvm/lib/CASObjectFormats/SimpleV1.cpp

```diff
@@ -0,0 +1,1217 @@
+
+    return Error::success();
+  }
+
+  unsigned ContentSize;
+  E = encoding::consumeVBR8(Remaining, ContentSize);
+  if (E)
+    return E;
+  auto Data = consumeDataOfSize(Remaining, ContentSize);
+  if (!Data)
+    return Data.takeError();
+
+  ArrayRef<char> Content(Data->data(), ContentSize);
+  auto &B = LGB.getLinkGraph()->createContentBlock(
+      *SectionInfo->Section, Content, Address, Alignment, AlignOffset);
+  if (auto E = LGB.addBlock(BlockIdx, &B))
+    return E;
+
+  // When not embedding, use Remaining to indicate if there is an EdgeList to
+  // decode.
+  unsigned RemainSize = NestEdgesInBlock ? Remaining.size() : 1;
+  LGB.setBlockRemaining(BlockIdx, RemainSize);
+  return Error::success();
+}
+
+Error BlockRef::materializeEdges(LinkGraphBuilder &LGB,
+                                 unsigned BlockIdx) const {
+  auto BlockInfo = LGB.getBlockInfo(BlockIdx);
+  if (!BlockInfo)
+    return BlockInfo.takeError();
+
+  // Nothing remains, no edges.
+  if (!BlockInfo->Remaining)
+    return Error::success();
+
+  if (!NestEdgesInBlock) {
+    auto Edge = LGB.getNode<EdgeListRef>(BlockInfo->BlockIdx++);
+    if (!Edge)
+      return Edge.takeError();
+
+    if (auto Err = Edge->materialize(LGB, *BlockInfo->Block))
+      return Err;
+
+    return Error::success();
+  }
+
+  auto Remaining = getData().take_back(BlockInfo->Remaining);
+  unsigned EdgeSize;
+  auto E = encoding::consumeVBR8(Remaining, EdgeSize);
+  if (E)
+    return E;
+
+  for (unsigned I = 0; I < EdgeSize; ++I) {
+    unsigned SymbolIdx;
+    E = encoding::consumeVBR8(Remaining, SymbolIdx);
+    if (E)
+      return E;
+
+    auto Symbol = LGB.getSymbol(SymbolIdx);
+    if (!Symbol)
+      return Symbol.takeError();
+
+    unsigned char Bits, Kind;
+    unsigned Offset, Addend, Relocation = 0;
+    E = encoding::consumeVBR8(Remaining, Bits);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Kind);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Offset);
+    if (!E)
+      E = encoding::consumeVBR8(Remaining, Addend);
+    if (E)
+      return E;
+
+    bool IsRelocation = Bits & (1U << 1);
+    if (IsRelocation)
+      E = encoding::consumeVBR8(Remaining, Relocation);
+    if (E)
+      return E;
+
+    BlockInfo->Block->addEdge((jitlink::Edge::Kind)Kind, Offset, **Symbol,
+                              Addend);
+  }
+
+  return Error::success();
+}
+
+Expected<BlockRef> BlockRef::get(Expected<ObjectFormatNodeRef> Ref) {
+  auto Specific = SpecificRefT::getSpecific(std::move(Ref));
+  if (!Specific)
+    return Specific.takeError();
+
+  return BlockRef(*Specific);
+}
+
+Expected<SymbolRef> SymbolRef::create(CompileUnitBuilder &CUB,
+                                      const jitlink::Symbol &S) {
+  Expected<Builder> B = Builder::startNode(CUB.Schema, KindString);
+  if (!B)
+    return B.takeError();
+
+  auto Name = CUB.getOrCreateName(S.getName());
+  if (!Name)
+    return Name.takeError();
```

Comment:
~~It looks like `Name` is dropped on the floor... not seeing where it gets encoded?~~

I figured it out -- probably worth a comment if it stayed like this, but since SymbolRefs will be inlined into the CompileUnit that doesn't seem important.